pkg/state/proof.go

```diff
@@ -0,0 +1,116 @@
+package state
+
+import (
+	"bytes"
+
+	"github.com/IBM-Blockchain/bcdb-server/pkg/crypto"
+	"github.com/IBM-Blockchain/bcdb-server/pkg/types"
+	"github.com/pkg/errors"
+)
+
+var DeleteBytes = []byte{1}
+
+type Proof struct {
+	// For each node in trie path, it contains bytes of all node fields and []byte{1} in case of deleted flag true
+	// Path is from rom node contains value to root node
+	// Exactly same byte slices used to calculate node hash.
+	Path []*types.MPTrieProofElement
+}
+
+func (p *Proof) Verify(leafHash, rootHash []byte, isDeleted bool) (bool, error) {
+	pathLen := len(p.Path)
+
+	if pathLen == 0 {
+		return false, errors.New("proof can't be empty")
+	}
+
+	// In case deleted value, node that contains it should contain []byte{1} between its hashes/bytes
+	if isDeleted {
+		isDeleteFound := false
+		for _, hash := range p.Path[0].GetHashes() {
+			if bytes.Equal(hash, DeleteBytes) {
+				isDeleteFound = true
+				break
+			}
+		}
+		if !isDeleteFound {
+			return false, nil
+		}
+	}
+
+	hashToFind := leafHash
+
+	// Validation algorithm just checks is hashToFind (current node/value hash) is part of hashes/bytes
+	// list in node above. We start from value hash (leafHash) and continue to root stored in block
+	for i := 0; i < pathLen; i++ {
+		isHashFound := false
+		for _, hash := range p.Path[i].GetHashes() {
+			if bytes.Equal(hash, hashToFind) {
+				isHashFound = true
+				break
+			}
+		}
+		if !isHashFound {
+			return false, nil
+		}
+
+		var err error
+		// hash here calculated same way as node hash calculated
+		hashToFind, err = CalcHash(p.Path[i].GetHashes())
+		if err != nil {
+			return false, err
+		}
+	}
+	// Check if calculated root hash if equal to supplied (stored in block)
+	return bytes.Equal(rootHash, hashToFind), nil
+}
+
+func (p *Proof) GetPath() []*types.MPTrieProofElement {
+	return p.Path
+}
```

Comment:
Why is this needed if Path is exported? 
maybe Path -> path?

Reply:
Let me check in sdk